feat: enable additional metadata collection (under feature flag)

GavinZZ opened the pull request:
### Issue # (if applicable)

Closes https://github.com/aws/aws-cdk/issues/33260

Relevant discussion https://github.com/aws/aws-cdk/discussions/33198

### Note

The majority of the code changes are auto generated so you'll see hundreds of `addConstructMetadata` method call across different L2 resources.

This method comes from this change https://github.com/aws/aws-cdk/pull/33232/files#diff-81f821b1205e7040fc3103bf7c0114060a6d5c43ebd2994aa4ed5906e42c9c5fR33. The main code change that needs to be reviewed is in `packages/aws-cdk-lib/core` as well as `tools/@aws-cdk/construct-metadata-updater`

### Reason for this change

This discussion aims to expand the scope of usage data collected by the AWS CDK to better inform CDK development and improve communication for customer-impacting topics. Currently, for those that opt in, the CDK collects usage data on your CDK version and which L2 constructs you use.

### Description of changes

1. Update CDK synthesis code to additionally handle resource metadata. On feature flag set to true, synthesis will not only inject Metadata usage like version and construct name, it will additionally look for any construct/method/feature flag metadata injected during resource creation. On feature flag set to false, it should be the same as before.
2. One-time tool metadata-updater to automatically find the right classes and add import statements and add metadata statements. The tool can be run multiple times and should not add additional import or add metadata statements to files that already been added. An action item is to link the tool to a GHA to periodically run this.
3. Build a workflow (that will be linked to GHA in the future) and when redacting, redact based on the value. The workflow on run will parse all files in aws-cdk repository and built a JSON file that contains all constructs and loggable properties of the construct. When redacting, only log the property key if the key exists in the JSON file. The value will be logged only if the value is not a * in the JSON file. Everything else is redacted for safely.
4. Build a JSON blueprint of the ENUM values and do not redact ENUM values.

Consider the following example

```
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';

class MyStack extends cdk.Stack {
  constructor(scope, id, props) {
    super(scope, id, props);

    // Create an S3 bucket (L2 construct)
    const myBucket = new s3.Bucket(this, 'MyBucket', {
      bucketName: 'my-cdk-example-bucket', // String type
      versioned: true,                    // Boolean type
      removalPolicy: cdk.RemovalPolicy.DESTROY, // ENUM type
      lifecycleRules: [{                  // Array of object type
        expirationDate: new Date('2019-10-01'),
        objectSizeLessThan: 600,
        objectSizeGreaterThan: 500,
      }],
    });

    // Use a method of the L2 construct to define additional properties
    myBucket.addLifecycleRule({
      id: 'ExpireOldObjects',
      enabled: true, // Boolean
      expiration: cdk.Duration.days(90), // Expire objects after 90 days
    });
  }
}

// Define the CDK app and stack
const app = new cdk.App();
new MyStack(app, 'MyStack');
app.synth();
```

At synthesis, usage data is collected, compressed, and stored in the `AWS::CDK::Metadata resource`. Based on **current** behavior, the following is an example of the usage data that will be collected from our example app:
```
{ 
    "fqn": "aws-cdk-lib.aws-s3.Bucket", 
    "version": "v2.170.0" 
}
```

With this proposal, the following usage data will be collected. The * value replaces property values that will be redacted from data collection: 
```
{ 
    "fqn": "aws-cdk-lib.aws_s3.Bucket", 
    "version": "2.170.0", 
    "metadata": [ 
        { 
            "type": "aws:cdk:analytics:construct", 
            "data": { 
                "bucketName": "*", 
                "versioned": true, 
                "removalPolicy": "cdk.RemovalPolicy.DESTROY", 
                "lifecycleRules": [ 
                    { 
                        "expirationDate": "*", 
                        "objectSizeLessThan": "*", 
                        "objectSizeGreaterThan": "*" 
                    } 
                ] 
            } 
        }, 
        { 
            "type": "aws:cdk:analytics:method", 
            "data": { 
                "name": "addLifecycleRule", 
                "prop": { 
                    "id": "*", 
                    "enabled": true, 
                    "expiration": "*",
                } 
            } 
        } 
    ] 
}
```

### Describe any new or updated permissions being added

No

### Description of how you validated changes

Many new unit tests added to verify different behaviour of various functions and methods introduced. One integ test file is added to test the deployability.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ff947cf9a21102abc7da7c9027e70593c3512a96
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a094dfd7-4e31-4a16-9350-61f07589e583%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

GavinZZ commented on packages/@aws-cdk/aws-amplify-alpha/lib/app.ts:
```diff
@@ -224,6 +225,8 @@ export class App extends Resource implements IApp, iam.IGrantable {
 
   constructor(scope: Construct, id: string, props: AppProps) {
     super(scope, id);
+    // Enhanced CDK Analytics Telemetry
+    addConstructMetadata(this, props);
```
The majority of the code changes are auto generated so you'll see hundreds of `addConstructMetadata` method call across different L2 resources.

This method comes from this change https://github.com/aws/aws-cdk/pull/33232/files#diff-81f821b1205e7040fc3103bf7c0114060a6d5c43ebd2994aa4ed5906e42c9c5fR33. The main code change that needs to be reviewed is in `packages/aws-cdk-lib/core` as well as `tools/@aws-cdk/construct-metadata-updater`

moelasmar replied:
We only support collecting metadata from the constructor for now .. right ?

GavinZZ replied:
This PR only supports constructor collection. I'll work on the method collections but I don't think that blocks us reviewing and merging this PR.

moelasmar commented on packages/aws-cdk-lib/core/lib/metadata-resource.ts:
```diff
@@ -0,0 +1,174 @@
+  try {
+    addMetadata(scope, MetadataType.CONSTRUCT, props);
+  } catch (e) {
+    /**
+     * Errors are ignored here.
+     * We do not want the metadata parsing to block users to synth or
+     * deploy their CDK application.
+     *
+     * Without this, it will just fall back to the previous metadata
+     * collection strategy.
+     */
```
should we add some logging to allow us to investigate if there is any issue ?

GavinZZ replied:
Added a warning statement for debuggging purposes

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f03fadc893fc2368511f3e8e1817f97c3b79236d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eee577ee-ef06-4a85-be78-811f515f2b9d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

moelasmar commented on tools/@aws-cdk/construct-metadata-updater/lib/metadata-updater.ts:
```diff
@@ -0,0 +1,500 @@
+  "decdk",
+  "awslint",
+  "test",
+  "aws-eks-v2-alpha"
```
why do you skip eks v2, and if this is a temp change, can you add TODO comment here, and also in the EKS alpha module to let us know when we should remove this line

GavinZZ replied:
When I made the changes, eks-v2-alpha was marked as private. It seems that it's now public so we should be able to add it now. Let me remove this check.

moelasmar commented:
are you going to add the GH workflow in a separate PR ?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 644a5d9e6ce82647521ec4e586856cdb13e43cc3
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1bab782b-8fc5-46a1-816f-91358636415d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

GavinZZ commented:
> are you going to add the GH workflow in a separate PR ?

Yes, it's going to be a separate PR.

moelasmar commented on packages/aws-cdk-lib/core/lib/analytics-data-source/enums.ts:
```diff
@@ -0,0 +1,4028 @@
+/* eslint-disable quote-props */
+/* eslint-disable @stylistic/comma-dangle */
+/*
+ * Do not edit this file manually. To prevent misconfiguration, this file
+ * should only be modified by an automated GitHub workflow, that ensures
+ * that the ENUMs present in this list
```
instead of depend on good intensions, can you update our pr-linter check to make sure that any change to these files are done by a commit from `aws-cdk-automation`, and reject any PR that violate that

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 445111348e57bb321c8e80887912258b41c2831d
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0e73c319-51c4-4731-8984-6c66f50c41b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

xazhao reviewed (commented):
Looks good to me overall.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: c281e840977ae32647f693ffeacef878412039a0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6ea3d007-0f2f-414f-b68f-4136d33e0214%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

GavinZZ commented on packages/aws-cdk-lib/core/lib/metadata-resource.ts:
```diff
@@ -0,0 +1,176 @@
+ */
+export function redactMetadata(fqn: string, data: any): any {
+  // A valid fqn is consists of 3 parts, i.e. `aws-cdk-lib.aws-lambda.Function`.
+  const fqnParts = fqn.replace(/[-_]/g, '-').split('.');
```
Yes, for example, IAM's `fqn` is something like `"fqn": "aws-cdk-lib.aws_iam.CfnRole",`

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 178289910a4848cd7f6531e3c9e6d37ef2b3d342
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ed2df0f3-8e72-4cd7-9cfd-f4539bb4ba2a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

GavinZZ commented:
Added `exempt-codecov` because we added `addConstructMetadata` statements to a number of L2 constructs. We can't exhaustively test all changes and the code coverage drop is expected.

GavinZZ commented:
@mergify update

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1451998f4deee500c99b04bbb7e20290d46fd92f
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a27f6e0-fb88-4020-8ac8-744892a89f8a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
